src/core/monitoring.cpp

```diff
@@ -12,19 +12,234 @@
  *   Alois Zoitl - initial implementation and rework communication infrastructure
  *   Martin Melik Merkumians - adds functionality for W/CHAR
  *******************************************************************************/
+#include <algorithm>
 #include <forte_config.h>
 #include "monitoring.h"
+#include "funcbloc.h"
 #include "resource.h"
 #include "device.h"
 #include "ecet.h"
 #include "core/util/string_utils.h"
 #include "../arch/timerha.h"
+#include "datatypes/forte_array.h"
+#include "datatypes/forte_struct.h"
 
 using namespace std::string_literals;
 using namespace forte::core;
+using namespace forte::core::internal;
 
 const std::string cgClosingXMLTag = "\">"s;
 
+namespace {
+  constexpr auto watchEntryComperator = [](const CWatchEntry &paItem, CStringDictionary::TStringId paPortId) {
+    return paItem.getPortId() < paPortId;
+  };
+
+  constexpr auto monitoringEntryComperator = [](const SFBMonitoringEntry &paItem, const CFunctionBlock *const paFB) {
+    return &paItem.getFB() < paFB;
+  };
+
+  void addDataWatch(internal::SFBMonitoringEntry &paFBMonitoringEntry,
+                    CStringDictionary::TStringId paPortId,
+                    CIEC_ANY &paDataVal) {
+    auto &dataWatches = paFBMonitoringEntry.mWatchedDataPoints;
+    auto it = std::lower_bound(dataWatches.begin(), dataWatches.end(), paPortId, watchEntryComperator);
+
+    if (it != dataWatches.end() && it->getPortId() == paPortId) {
+      // the data point is already in the watch list
+      return;
+    }
+
+    dataWatches.emplace(it,
+                        CDataWatchEntry(paPortId, paDataVal, paFBMonitoringEntry.getFB().getAbsDataPortNum(paPortId)));
```

Comment:
I would suggest to either use `insert` when there is already a (temporary) instance or supply the individual arguments to actually construct the element in-place with `emplace`:
```suggestion
    dataWatches.emplace(it, paPortId, paDataVal, paFBMonitoringEntry.getFB().getAbsDataPortNum(paPortId));
```

Reply:
done, didn't know this works.